### PR TITLE
#271 add go support

### LINKED ISSUE #271: Add support for Go projects
PK must support
* Generating the list of updated dependencies based on `go.mod` files

Note: Licenses for test dependencies are missing, see https://github.com/exasol/project-keeper/issues/279

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #271

### CODE REVIEW

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package com.exasol.projectkeeper;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.logging.Logger;
+
+public class ProjectKeeperLauncher {
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:7]: Missing javadoc

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    private static final java.util.logging.Logger LOGGER = Logger.getLogger(ProjectKeeperLauncher.class.getName());
+
+    public static void main(final String[] args) {
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:10]: missing javadoc

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        if (!success) {
+            throw new IllegalStateException("invalid");
+        }
+    }
+
+    private static class JULLogger implements com.exasol.projectkeeper.Logger {
+
+        @Override
+        public void info(final String message) {
+            System.out.println(message);
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:23]:
```suggestion
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            LOGGER.info(message);
+        }
+
+        @Override
+        public void warn(final String message) {
+            System.out.println(message);
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:29]:
```suggestion
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+            LOGGER.warning(message);
+        }
+
+        @Override
+        public void error(final String message) {
+            System.out.println(message);
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:35]:
```suggestion
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import com.exasol.projectkeeper.config.ProjectKeeperConfig.Source;
+import com.exasol.projectkeeper.config.ProjectKeeperConfig.SourceType;
+import com.exasol.projectkeeper.shared.dependencies.ProjectDependencies;
+import com.exasol.projectkeeper.shared.dependencychanges.DependencyChangeReport;
+import com.exasol.projectkeeper.sources.AnalyzedGolangSource;
+import com.exasol.projectkeeper.sources.AnalyzedSource;
+import com.exasol.projectkeeper.sources.analyze.golang.GolangServices;
+import com.exasol.projectkeeper.sources.analyze.golang.ModuleInfo;
+
+public class GolangSourceAnalyzer implements LanguageSpecificSourceAnalyzer {
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java:20]: Missing Javadoc.

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    private final GolangServices golangServices;
+
+    public GolangSourceAnalyzer(final ProjectKeeperConfig config) {
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java:24]: Misssing javadoc

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        return sources.stream().map(source -> analyzeSource(projectDir, source)).collect(toList());
+    }
+
+    private AnalyzedSource analyzeSource(final Path projectDir, final Source source) {
+        if (source.getType() != SourceType.GOLANG) {
+            throw new IllegalStateException(ExaError.messageBuilder("F-PK-CORE-130")
+                    .message("Analyzing of {{type}} is not supported by GolangSourceAnalyzer", source.getType())
+                    .ticketMitigation().toString());
+        }
+
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/GolangSourceAnalyzer.java:43]:
```suggestion
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/LanguageSpecificSourceAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.exasol.projectkeeper.sources.analyze;
+
+import java.nio.file.Path;
+import java.util.List;
+
+import com.exasol.projectkeeper.config.ProjectKeeperConfig.Source;
+import com.exasol.projectkeeper.sources.AnalyzedSource;
+
+public interface LanguageSpecificSourceAnalyzer {
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/LanguageSpecificSourceAnalyzer.java:9]: Missing javadoc

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/LanguageSpecificSourceAnalyzer.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+    List<AnalyzedSource> analyze(Path projectDir, List<Source> sources);
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/LanguageSpecificSourceAnalyzer.java:11]: Missing javadoc

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/MavenSourceAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
+package com.exasol.projectkeeper.sources.analyze;
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/MavenSourceAnalyzer.java:1]: Reviewed until here.

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplatesFactory.java [hunk 12]
```diff
@@ -75,4 +87,8 @@ private boolean isMvnRootProject(final AnalyzedSource source) {
         }
         return templates;
     }
+
+    private List<FileTemplate> getGolangTemplates() {
+        return emptyList();
```
Comment (jakobbraun) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplatesFactory.java:92]: Can you create an issue for adding CI build tables for GO?

File: project-keeper/src/test/java/com/exasol/projectkeeper/GolangProjectFixture.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        Files.writeString(this.projectDir.resolve("go.mod"), content);
+    }
+
+    private void writeMainGoFile() throws IOException {
+        final String content = "package main\n" + "import (\n" + "    \"github.com/exasol/exasol-driver-go\"\n" + ")\n"
+                + "func main() {\n" + "    exasol.NewConfig(\"sys\", \"exasol\")\n" + "}\n";
+        Files.writeString(this.projectDir.resolve("main.go"), content);
+    }
+
+    private void executeSplit(final String command) {
```
Comment (jakobbraun) [inline project-keeper/src/test/java/com/exasol/projectkeeper/GolangProjectFixture.java:73]: That method name is not very clear.

File: project-keeper/src/test/resources/go.mod [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/sirupsen/logrus v1.8.1 // indirect
+	go.opencensus.io v0.23.0 // indirect
+	golang.org/x/net v0.0.0-20220325170049-de3da57026de // indirect
+	golang.org/x/sys v0.0.0-20220330033206-e17cdc41300f // indirect
+	google.golang.org/genproto v0.0.0-20220329172620-7be39ac1afc7 // indirect
+	google.golang.org/grpc v1.45.0 // indirect
+	google.golang.org/protobuf v1.28.0 // indirect
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
+)
```
Comment (jakobbraun) [inline project-keeper/src/test/resources/go.mod:46]: do you really need such a big file for testing? Or do you use it as a real-live smoke test?

File: project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.logging.Logger;
+
+import com.exasol.errorreporting.ExaError;
+
+/**
+ * This is the main entry point for launching Project Keeper on the command line.
+ */
+public class ProjectKeeperLauncher {
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/ProjectKeeperLauncher.java:13]: 👍

File: project-keeper/src/test/resources/go.mod (hunk 14 above)
Comment (kaklakariada) [reply project-keeper/src/test/resources/go.mod:46]: I use it only in a single test case. All other test cases use a minimal inline string.

File: project-keeper/src/test/java/com/exasol/projectkeeper/config/ProjectKeeperConfigWriter.java [hunk 16]
```diff
@@ -31,14 +33,19 @@ public ConfigForWriting prepareWriting(final ProjectKeeperConfig config) {
             sourcesForWriting
                     .add(new ConfigForWriting.Source(source.getPath().toString(), source.getType().name(), modules));
         }
-        return new ConfigForWriting(sourcesForWriting, config.getLinkReplacements(), config.getExcludes());
+        String version = null;
+        if (config.getVersionConfig() instanceof FixedVersion) {
+            version = ((FixedVersion) config.getVersionConfig()).getVersion();
+        }
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/test/java/com/exasol/projectkeeper/config/ProjectKeeperConfigWriter.java:39]: ✅

File: project-keeper/src/test/java/com/exasol/projectkeeper/GolangProjectFixture.java (hunk 13 above)
Comment (kaklakariada) [reply project-keeper/src/test/java/com/exasol/projectkeeper/GolangProjectFixture.java:73]: ✅

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/changesfile/DependencySectionFixer.java [hunk 17]
```diff
@@ -48,6 +47,9 @@ private NamedDependencyChangeReport getDependencyChangesOfSource(final AnalyzedS
         if (source instanceof AnalyzedMavenSource) {
             final AnalyzedMavenSource mvnSource = (AnalyzedMavenSource) source;
             return new NamedDependencyChangeReport(mvnSource.getProjectName(), mvnSource.getDependencyChanges());
+        } else if (source instanceof AnalyzedGolangSource) {
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/validators/changesfile/DependencySectionFixer.java:50]: ✅

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/dependencies/DependenciesValidator.java [hunk 18]
```diff
@@ -67,6 +66,9 @@ private ProjectWithDependencies getDependencies(final AnalyzedSource source) {
             final String projectName = mvnSource.getProjectName();
             final List<ProjectDependency> dependencies = mvnSource.getDependencies().getDependencies();
             return new ProjectWithDependencies(projectName, dependencies);
+        } else if (source instanceof AnalyzedGolangSource) {
+            final AnalyzedGolangSource goSource = (AnalyzedGolangSource) source;
+            return new ProjectWithDependencies(goSource.getProjectName(), goSource.getDependencies().getDependencies());
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/validators/dependencies/DependenciesValidator.java:71]: ✅

File: project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/golang/GolangDependencyLicense.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
+package com.exasol.projectkeeper.sources.analyze.golang;
+
+import com.exasol.projectkeeper.shared.dependencies.License;
+
+class GolangDependencyLicense {
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/sources/analyze/golang/GolangDependencyLicense.java:5]: ✅

File: project-keeper/src/main/java/com/exasol/projectkeeper/LicenseNameReader.java [hunk 20]
```diff
@@ -30,7 +30,7 @@ private String readLicenseFile(final Path projectDir) {
         try {
             return Files.readString(projectDir.resolve("LICENSE"));
         } catch (final IOException exception) {
-            throw new IllegalStateException(ExaError.messageBuilder("F-PK-CORE-124").message(
+            throw new IllegalStateException(ExaError.messageBuilder("F-PK-CORE-139").message(
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/LicenseNameReader.java]: ✅

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**